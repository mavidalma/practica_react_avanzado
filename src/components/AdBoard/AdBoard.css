
.ads-wall{
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-evenly;
    align-items: center;
    padding: 10px;
}
.card-container {
    margin: 20px auto;
    border: 2px solid black;
    overflow: hidden;
}
.card-container .ad {
    display: flex;
    flex-direction: row;
    width: 600px;
    max-height: 300px;
    padding: 10px;

}
.info {
    width: 60%;
}
.image-container {
    width: 40%;
}

.image {
    max-width: 100%;
}
