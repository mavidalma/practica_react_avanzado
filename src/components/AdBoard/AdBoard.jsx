import React, { Component } from 'react';
import { fetchAds, getTags } from '../../api_caller';
import './AdBoard.css';
import { BrowserRouter as Router, Link } from "react-router-dom";
import AdFilter from './AdFilter/AdFilter';
import AdCard from '../AdCard/AdCard';


export default class AdBoard extends Component {

    constructor(props) {
        super(props);
        this.state = {
            error: "",
            data: [],
            tags: [],
            success: true,
            maxPrice: 100000000,
        }
    }

    getAds = async(query) => {
       await fetchAds(query)
            .then(data => {
                this.setState({success: data.success});
                this.setState({error: data.error});
                this.setState({data: data.results})
            });
      //  this.getMaxPrice(this.state.data);
    }

    getTags = async () => {
        await getTags()
            .then(data => this.setState({ tags: data }));
    }

    getMaxPrice = (data) => {
        const topPrice = data.map(item => item.price)
            .reduce((previous, current) => (current > previous) ? current : previous);
        this.setState({
            maxPrice: topPrice,
        })
    }

    componentDidMount() {
        this.getAds("");
        this.getTags();
        console.log("componentdidmount") //to check out on console the number of times the component remounts and why it renders empty before mounting
    }

    render() {

        if(this.state.success){
        return (
            <div>
                <AdFilter data={this.state.data}
                    getAds={this.getAds}
                    tags={this.state.tags}
                    maxPrice = {this.state.maxPrice}
                    props={this.props}
                />
                <div className="ads-wall">{this.state.data.map(card => {
                    return (
                        < div key={card._id} className="card-container">
                            <AdCard data={card} />
                        </div>
                    )
                })}
                </div>
            </div>
        )
    } else { //I tried to use a conditional (this.state.error === "Error: not logged in") {} but since it first renders an empty version of the app, it doesn work
        return (
                <div className="error-message">
                  <h1>Please log in</h1>
                  <h2> In order to use our patform you have to be a registered user.</h2>
                    
                  <p>Already register? <Link to="/login"><button>Go to login</button> </Link> </p>
                  <p>Want to create an account? <Link to="/register"><button>Go to register</button> </Link></p>
                </div>
              )
    }
    }
}