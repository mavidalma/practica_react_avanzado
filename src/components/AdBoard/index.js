import { connect } from 'react-redux';

import AdBoard from './AdBoard';

import { fetchAds, fetchTags, getUserFromStorage } from '../../store/actions';
import { getAds, getTags, isLogged } from '../../store/selectors';

function mapStateToProps(state, ownProps) {
  return {
    ads: getAds(state),
    tags: getTags(state),
    user: isLogged(state) 
  }
}

const mapDispatchToProps = {
    fetchAds,
    fetchTags,
    getUserFromStorage

}

export default connect(mapStateToProps, mapDispatchToProps)(AdBoard);