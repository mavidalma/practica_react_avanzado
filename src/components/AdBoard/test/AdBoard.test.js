import React from 'react';
import AdBoard from '../AdBoard';
import { shallow } from 'enzyme';

describe('snapshop test od AdBoard', ()=> {
    const props = {
        ads: [
        {
            name: "test1",
            description: "description1",
            price: 25,
            type: 'sell',
            tag: 'motor',
        },
        {
            name: "test2",
            description: "description2",
            price: 10,
            type: 'buy',
            tag: "work"
        },
        ],
        loading: false,
        tags: ['motor', 'lifestyle', 'work'],
        user: true,
        fetchAds: jest.fn(),
        getUserFromStorage: jest.fn(),
        fetchTags: jest.fn(),
        className: 'test',
    }
    let wrapper;

    beforeEach(()=> {
        wrapper = shallow(<AdBoard {...props} />);
    });

    test('AdBoad snapshot', ()=> {
        expect(wrapper).toMatchSnapshot();
    });

    test("component renders", ()=> {
        expect(wrapper.exists()).toBe(true)
    });
    test("has tags props", ()=> {
        expect(wrapper.props().tags).toEqual(props.tags)
    });
    
/*
    test('should render Loading component when isLoading is set to true', ()=> {
        wrapper.setProps({ loading: true });
        console.log(wrapper.debug());
        expect(wrapper.find('Loading')).toBe(true);

    });*/
/*
    test("Adwall renders when component is loaded", ()=> {
        console.log(wrapper.debug());
        expect(wrapper.find('AdWall')).toBe(true);
    })*/
})