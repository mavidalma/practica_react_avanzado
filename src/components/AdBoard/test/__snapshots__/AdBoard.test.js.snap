// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshop test od AdBoard AdBoad snapshot 1`] = `
<div>
  <Connect(AdFilter)
    maxPrice={100000000}
    props={
      Object {
        "ads": Array [
          Object {
            "description": "description1",
            "name": "test1",
            "price": 25,
            "tag": "motor",
            "type": "sell",
          },
          Object {
            "description": "description2",
            "name": "test2",
            "price": 10,
            "tag": "work",
            "type": "buy",
          },
        ],
        "className": "test",
        "fetchAds": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "fetchTags": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "getUserFromStorage": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
        "loading": false,
        "tags": Array [
          "motor",
          "lifestyle",
          "work",
        ],
        "user": true,
      }
    }
    tags={
      Array [
        "motor",
        "lifestyle",
        "work",
      ]
    }
  />
  <AdWall
    ads={
      Array [
        Object {
          "description": "description1",
          "name": "test1",
          "price": 25,
          "tag": "motor",
          "type": "sell",
        },
        Object {
          "description": "description2",
          "name": "test2",
          "price": 10,
          "tag": "work",
          "type": "buy",
        },
      ]
    }
  />
</div>
`;
